### DISCUSSION TITLE
Which is the URL for the metrics?

### DISCUSSION THREAD

**Discussion by juangascon** [kube-green/kube-green#265] [Q&A] [answered]
Hello.
I have read all the documentation and I am not able to know which is the IP of the URL to get the metrics.
I suppose that the URL would be http://<IP>:8080/metrics but I do not know if the IP is the cluster endpoint.
Thanks for your help.

**Comment by davidebianchi** [comment, marked as answer]
The metrics endpoint is exposed from the manager itself. So, it is possible to point to the `kube-green` service (so an url like `http://kube-green:8080/metrics`) if you are inside the cluster.

**Reply by juangascon** [reply]
Hello.
Thanks for your answer. Unluckily I am not a Kubernetes expert and I do not really understand your answer :-(
How do I get inside the cluster?
When I type `kubectl -n kube-green get services` I can see the service (there are 2 indeed) with their private IPs. 
But I would like to show these metrics in a web-browser and I only know to do that with external IPs
I do not arrive to expose the deployment externally. There is only one deployment and when I expose it, the endpoint is not present :-(
Should I deploy a Grafana to get the kind of charts we can see in the 'Real Use case' docs?

**Reply by juangascon** [reply]
I have used  `kubectl proxy` and `kubectl port-forward` but I do not know how to reroute.
I can see 2 services:
- kube-green-controller-manager-metrics-service on port 8443
- kube-green-webhook-service on port 443
Looking into the deployment YAML, I understand that the 443 is rerouted to the 9443. I have created the FW rule to open that port in GCP but nothing is shown
If I run the kubectl proxy on port 8000 to not interfere with the 8080, I get nothing but cluster internal info.
So, I am blocked to get those metrics.
Besides that, the rules run perfectly and I can observe that the pods are sleeping with `kubectl get pods`.

**Reply by davidebianchi** [reply]
Hello @juangascon!
The charts in the documentation are taken from Grafana which takes the data from a Prometheus. The data in the chart are made from the CPU and memory of the pods inside the namespaces of the cluster.
Those are not metrics data exposed by kube-green.

It is possible to show the metrics exposed by kube-green (which should be collected by prometheus) with port-forward:

```sh
kubectl port-forward NAME_OF_THE_KUBE_GREEN_POD 8080
```

And, once the connection is estabilished:

```sh
curl http://localhost:8080/metrics
```

But those metrics are only the metrics of kube-green, and are not useful to have the chart in the documentation.

**Reply by juangascon** [reply]
Hello @davidebianchi 
Thanks a lot for your answer. I had not understood that I have to install and use Prometheus to get the metrics from the pods.
I thought that kube-green would expose them as json files that I could inject in a time-based database to see later in a data visualizer.
I understand that the metics I would get with kube-green are not those of the pods but kube-green itself
In any case, your comment will help me to get metrics in clusters.

**Reply by davidebianchi** [reply]
Nice! Can we set this as the answer?

**Reply by juangascon** [reply]
@davidebianchi OK. I think this could be the answer.
I have tried again the tutorial and the commands above.
I got a return with a bunch of fields and info that I suppose are useful for Prometheus.
Among all those fields, I found this:
```
# HELP kube_green_current_sleepinfo Info about SleepInfo resource
# TYPE kube_green_current_sleepinfo gauge
kube_green_current_sleepinfo{name="sleep-test",namespace="sort-list"} 1
```
I suppose that this is what is supposed we would retrieve and Prometheus will show it with a gauge put at 1

**Reply by davidebianchi** [reply]
Yes, correct. This metrics is the number of SleepInfo configured in the cluster. At the moment, it is the only metrics: https://kube-green.dev/docs/metrics/

**Comment by juangascon** [comment]
I have followed the Tutorial "[Usage with kind](https://kube-green.dev/docs/tutorials/kind/)" and it works perfectly even in Cloud managed kubernetes. Though, I think it would be nice to show the metrics in this tutorial.
Would it possible to add in the tutorial how to get and display the metrics in a browser?
Thanks in advance

**Reply by davidebianchi** [reply]
I think it's out of scope to add how to setup the monitoring stack in the tutorial, it depends on what do you want to use to gather and view metrics and it is not related to kube-green.

To start, you can see how to setup the monitoring stack with the [Prometheus Operator](https://github.com/prometheus-operator/prometheus-operator) and the [Grafana chart repository](https://github.com/grafana/helm-charts/blob/main/charts/grafana/README.md). The charts showed are in the dashboards of the [monitoring mixin](https://monitoring.mixins.dev/).

In the future, for sure there will be some ready to use Grafana dashboards with the metrics exposed by kube-green, but at the moment I haven't created it yet (there is only 1 metric exposed by kube-green). Also, in the future there could be some frontend to manage and monitor the kube-green state, which could be added for sure in the tutorial.
